### PR TITLE
impl(GCS+gRPC): `WriterConnectionFinalized`

### PR DESCRIPTION
**Pull request by coryan**
Sometimes the application may try to resume an upload that was already finalized. This introduces an implementation of the `storage_experimental::WriterConnection` interface that is suited for that case.

Part of the work for #9134

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13023)

### CODE REVIEW

File: google/cloud/storage/internal/async/writer_connection_finalized_test.cc [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+TEST(AsyncWriterConnectionFinalized, Basic) {
+  auto expected = MakeTestObject();
+  AsyncWriterConnectionFinalized tested("test-upload-id", expected);
+
+  EXPECT_EQ(tested.UploadId(), "test-upload-id");
+  auto const state = tested.PersistedState();
+  ASSERT_TRUE(absl::holds_alternative<storage::ObjectMetadata>(state));
+  EXPECT_EQ(absl::get<storage::ObjectMetadata>(state), expected);
```
Comment (dbolduc) [inline google/cloud/storage/internal/async/writer_connection_finalized_test.cc:49]:
```cc
EXPECT_THAT(tested.PersistedState(), VariantWith<storage::ObjectMetadata>(expected));
```

Comment (coryan) [reply google/cloud/storage/internal/async/writer_connection_finalized_test.cc]: Fixed.

File: google/cloud/storage/internal/async/writer_connection_finalized.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * Implement the `AsyncWriterConnection` interface for resumed, but finalized,
+ * uploads.
+ *
+ * Applications may resume an upload that was already finalized. For example,
+ * the application may resume all pending uploads when it starts, using some
+ * kind of database to keep the pending uploads. Such an application may
+ * finalize an upload, crash or be terminated before having an opportunity to
+ * update its database, and then try to resume the upload only to discover it
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/async/writer_connection_finalized.h]: Changed to use bullet points.  Let me know what you think.

File: google/cloud/storage/internal/async/writer_connection_finalized.h [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ * is finalized.
+ *
+ * In this case we want to return an implementation of `AsyncWriterConnection`
+ * that contains the result of a finalized upload, but cannot
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/async/writer_connection_finalized.h]: Fixed.

**Review by dbolduc — APPROVED** [review]

**Pull request merged**